3-UsingAPIs/Week1/homework/ex3-rollDice.js

```diff
@@ -8,48 +8,43 @@
 - Does the problem described above still occur? If not, what would be your
   explanation? Add your answer as a comment to be bottom of the file.
 ------------------------------------------------------------------------------*/
-
 // TODO Remove callback and return a promise
-function rollDice(callback) {
-  // Compute a random number of rolls (3-10) that the dice MUST complete
-  const randomRollsToDo = Math.floor(Math.random() * 8) + 3;
-  console.log(`Dice scheduled for ${randomRollsToDo} rolls...`);
-
-  const rollOnce = (roll) => {
-    // Compute a random dice value for the current roll
-    const value = Math.floor(Math.random() * 6) + 1;
-    console.log(`Dice value is now: ${value}`);
-
-    // Use callback to notify that the dice rolled off the table after 6 rolls
-    if (roll > 6) {
-      // TODO replace "error" callback
-      callback(new Error('Oops... Dice rolled off the table.'));
-    }
-
-    // Use callback to communicate the final dice value once finished rolling
-    if (roll === randomRollsToDo) {
-      // TODO replace "success" callback
-      callback(null, value);
-    }
-
-    // Schedule the next roll todo until no more rolls to do
-    if (roll < randomRollsToDo) {
-      setTimeout(() => rollOnce(roll + 1), 500);
-    }
-  };
-
-  // Start the initial roll
-  rollOnce(1);
+function rollDice() {
+  return new Promise((resolve, reject) => {
+    const randomRollsToDo = Math.floor(Math.random() * 8) + 3;
+    console.log(`Dice scheduled for ${randomRollsToDo} rolls...`);
+
+    const rollOnce = (roll) => {
+      const value = Math.floor(Math.random() * 6) + 1;
+      console.log(`Dice value is now: ${value}`);
+
+      if (roll > 6) {
+        // TODO replace "error" callback
+        reject(new Error('Oops... Dice rolled off the table.'));
+      }
+      if (roll === randomRollsToDo) {
+        // TODO replace "success" callback
+        resolve(value);
+      }
+      if (roll < randomRollsToDo) {
+        setTimeout(() => rollOnce(roll + 1), 500);
+      }
+    };
+    // Start the initial roll
+    rollOnce(1);
+  });
 }
 
 // TODO Refactor to use promise
-rollDice((error, value) => {
-  if (error !== null) {
-    console.log(error.message);
-  } else {
-    console.log(`Success! Dice settled on ${value}.`);
-  }
-});
+rollDice()
+  .then((response) => console.log(`Success! Dice settled on ${response}`))
+  .catch((error) => console.log(error.message));
 
 // ! Do not change or remove the code below
 module.exports = rollDice;
+
+/*
+I noticed that when using promise if one error happened then it will not returns success at all;
+ but with callback it will returns success always for at least 1 success even if there're many errors.
+so it looks likes promise (AND gate &&) while callback is (OR gate ||)
```

Comment:
There is still the same behavior like callbacks. If `randomRollsToDo` is bigger than 7 then it will reject the promise in the 7th roll but continue rolling more until it reaches to the final amount of rolls.

The thing is that using `resolve` and `reject` doesn't exit the function like `return` and continues to execute sometimes unwanted code.

This is why a lot of times we will see `resolve`/`reject` paired with `return` statements to avoid issue like in this exercise